src/cli_utils.cc

```diff
@@ -78,70 +78,165 @@ Status LoadYAMLFile(const string& filename, YAML::Node &node) {
-            yaml << YAML::Key << "name";
-            yaml << YAML::Value << pr.first;
-            yaml << YAML::Key << "size";
-            yaml << YAML::Value << pr.second;
-            yaml << YAML::EndMap;
-        }
-        yaml << YAML::EndSeq;
+        os << "---" << endl;
+        YAML::Emitter yaml;
+        S(yaml_of_contigs(contigs, yaml));
+        os << yaml.c_str() << endl;
     }
 
     // Write alleles
-    yaml << YAML::Key << "alleles";
-    yaml << YAML::Value;
-    {
-        // Note: we might write out empty sites too
-        s = yaml_of_discovered_alleles(dsals, contigs, yaml);
-        if (s.bad()) return s;
+    for (auto &pr : dsals) {
+        os << "---" << endl;
+        YAML::Emitter yaml;
+        S(yaml_of_one_discovered_allele(pr.first, pr.second, contigs, yaml));
+        os << yaml.c_str() << endl;
     }
 
-    yaml << YAML::EndMap;
+    // special notation for end-of-file
+    os << "..." << endl;
 
     return Status::OK();
 }
 
-Status contigs_alleles_of_yaml(const YAML::Node& yaml,
-                               std::vector<std::pair<std::string,size_t> > &contigs,
-                               discovered_alleles &dsals) {
-    Status s;
-    if (!yaml.IsMap()) {
-        return Status::Invalid("not a map at top level");
+
+static string yaml_begin_doc = "---";
+static string yaml_end_doc_list = "...";
+
+static Status yaml_verify_begin_doc_list(std::istream &is) {
+    try {
+        string marker;
+        std::getline(is, marker);
+        if (marker != yaml_begin_doc)
+            return Status::Invalid("document does not start with `---`");
+        return Status::OK();
+    } catch (exception e) {
+        string err = e.what();
+        return Status::Failure("exception caught in yaml_verify_begin_doc_list: ", err);
     }
+}
+
+// In a YAML document built as a of document list, get the next document.
+//
+// Notes:
+// -  Catch any exceptions, and convert to bad status
+static Status yaml_get_next_document(std::istream &is,
+                                     YAML::Node &ans,
+                                     string& next_marker) {
+    try {
+        stringstream ss;
+
+        while (is.good() && !is.eof()) {
+            string line;
+            std::getline(is, line);
+
+            if (line.size() == 3) {
+                // check if we reached the end of this top level document
+                if (line == yaml_begin_doc ||
+                    line == yaml_end_doc_list) {
+                    // We have the entire document in memory. Convert to
+                    // a YAML node and return.
+                    next_marker = line;
+                    ans = std::move(YAML::Load(ss.str()));
+                    return Status::OK();
+                }
+            }
+
+            ss.write(line.c_str(), line.size());
+            ss.write("\n", 1);
+        }
+
+        return Status::Invalid("premature end of document, did not find End-of-Doc/End-of-File marker");
```

Comment:
It's alternatively possible at this point is that `!is.good()`, i.e. some I/O error occurred reading the stream.